Auto-detect Azure cloud name via IMDS

ungureanuvladvictor opened the pull request:
Via the instance metadata service we can query in which Azure cloud the
operator is running.

Signed-off-by: Vlad Ungureanu <vladu@palantir.com>

```release-note
Auto-detect Azure cloud name via IMDS
```

aanm commented on operator/provider_azure_flags.go:
```diff
@@ -26,9 +26,6 @@ import (
 
-	flags.String(operatorOption.AzureCloudName, "AzurePublicCloud", "Name of the Azure cloud being used")
-	option.BindEnvWithLegacyEnvFallback(operatorOption.AzureCloudName, "AZURE_CLOUD_NAME")
-
```
This changed was introduced in 1.9 so we shouldn't simply remove it.

One way to fix it is to mart the option as deprecated with `flags.MarkDeprecated()` and add it as part of the upgrade notes for 1.11 and simply state that this will be removed since it's auto-detected now.

ti-mo replied:
Alright, thank you!

ungureanuvladvictor replied:
Talked with @aanm on Slack and got a pointer on using `viper.IsSet` for checking if the cloud name CLI flag is passed in order to trigger the new behavior. This should allow the current users which override it to keep the behavior that they wanted and the users who do not override the flag to get the new auto-detection. Let me know if you folks agree with this.

errordeveloper commented on pkg/ipam/allocator/azure/azure.go:
```diff
@@ -48,6 +48,11 @@ func (*AllocatorAzure) Start(getterUpdater ipam.CiliumNodeGetterUpdater) (alloca
 
+	azureCloudName, err := azureAPI.GetAzureCloudName(context.TODO())
+	if err != nil {
+		return nil, fmt.Errorf("unsable to retrieve Azure cloud name: %w", err)
```
```suggestion
		return nil, fmt.Errorf("unable to retrieve Azure cloud name: %w", err)
```

errordeveloper commented on pkg/ipam/allocator/azure/azure.go:
```diff
@@ -48,6 +48,11 @@ func (*AllocatorAzure) Start(getterUpdater ipam.CiliumNodeGetterUpdater) (alloca
 
 	log.Info("Starting Azure IP allocator...")
 
+	azureCloudName, err := azureAPI.GetAzureCloudName(context.TODO())
```
I can see there is at least one other instance of `context.TODO()` in this function, do we have a plan to address these todos?

ungureanuvladvictor replied:
@errordeveloper -- I changed the signature of the `Allocator` interface to pass a context to the `Start`.

ungureanuvladvictor committed:
Add context.Context as param to the Start method of the AllocatorProvider

Signed-off-by: Vlad Ungureanu <vladu@palantir.com>

ungureanuvladvictor committed:
Auto-detect Azure cloud name via IMDS

Via the instance metadata service we can query in which Azure cloud the
operator is running.

Marked the "azure-cloud-name" flag as deprecated and to be removed in
1.11.

Signed-off-by: Vlad Ungureanu <vladu@palantir.com>

ti-mo commented on pkg/ipam/allocator/azure/azure.go:
```diff
@@ -19,6 +19,7 @@ import (
 	"fmt"
 
 	operatorMetrics "github.com/cilium/cilium/operator/metrics"
+	"github.com/cilium/cilium/operator/option"
```
This package is already imported on the subsequent line.

ti-mo commented on pkg/ipam/allocator/azure/azure.go:
```diff
@@ -48,10 +50,22 @@ func (*AllocatorAzure) Start(getterUpdater ipam.CiliumNodeGetterUpdater) (alloca
 
 	log.Info("Starting Azure IP allocator...")
 
+	var azureCloudName string
```
I think the following might be nicer:
```golang
azureCloudName := operatorOption.Config.AzureCloudName
if !viper.IsSet(operatorOption.AzureCloudName) {
    // Call IMDS...
}
```

ti-mo reviewed (approved):
Thanks Vlad! Left 2 small nits.
